lua/entities/gmod_wire_expression2/core/http_cl.lua

```diff
@@ -0,0 +1,157 @@
+lib.segmentBits = math.min(math.ceil(math.log(lib.segmentSize,2)) + 2,32)
+lib.sendInterval = 0.5-- seconds
+lib.HTTP_REQUEST_FAILED = 1024--enum for HTTP failure
+
+-- void function(string id, number interval, function func)
+-- Internal: Do not call.
+function lib.timerCoroutine(id,interval,func)
+	func = coroutine.wrap(func)
+	timer.Create(id,interval,0,function()
+		if func() then
+			timer.Remove(id)
+		end
+	end)
+end
+
+util.AddNetworkString(lib.netMsgID)
+
+lib.requests = {}
+
+-- void function(entity client, string url, function callback_success[, function callback_failure])
+function lib:request(client,url,callback_success,callback_failure)
+	-- When the client sends us the data from the request, we need to know which
+	-- request it was for, so a UID is used.
+
+	if(cvar_useClient:GetInt() ~= 0) then
+		local uid = self:newUID()
+
+		self.requests[uid] = {
+			success = callback_success,
+			failure = callback_failure,
+			ply = client
+		}
+
+		self:launchNewRequest(client,url,uid)
+	else
+		self.rawRequest(url,callback_success,callback_failure)
+	end
+end
+
+-- void function(entity client, string url, number uid)
+-- Internal: Do not call.
+function lib:launchNewRequest(client,url,uid)
+	net.Start(self.netMsgID)
+		net.WriteString(url)
+		net.WriteUInt(uid,8)
+	net.Send(client)
+end
+
+-- string, boolean function(void)
+-- Internal: Do not call.
+function lib:decodeRequestHeader()
+	return net.ReadUInt(8),net.ReadBool()
+end
+
+-- boolean function(void)
+-- Internal: Do not call.
+function lib:isSendingBody()
+	return net.ReadBool()
+end
+
+-- string function(void)
+-- Internal: Do not call.
+function lib:readSegment()
+	return net.ReadData(net.ReadUInt(self.segmentBits))
+end
+
+-- table, number function(void)
+-- Internal: Do not call.
+function lib:readMetadata()
+	return net.ReadTable(),net.ReadUInt(12)
+end
+
+-- table, number function(void)
+-- Internal: Do not call.
+function lib:readError()
+	return net.ReadString()
+end
+
+-- void function(void)
+-- Internal: Do not call.
+function lib:handleIncomingRequest(ply)
+	local uid,failure = self:decodeRequestHeader()
+
+	if self.requests[uid] then
+		local request = self.requests[uid]
+
+		if ply ~= request.ply then return end -- Only the player who was requested to do a HTTP request may send us information about it
+
+		if not failure then
+			request.body_compressed = request.body_compressed or "" -- define the compressed body if it does not already exist
+			local isSendingBody = self:isSendingBody()
+			if isSendingBody then
+				local segment = self:readSegment()
+				request.body_compressed = request.body_compressed..segment
+			else
+				local headers,code = self:readMetadata()
+
+				local body = ""
+				if request.body_compressed ~= "" then
+					body = util.Decompress(request.body_compressed)
+				end
+
+				if body then
+					request.success(body,#body,headers,code)
+				elseif request.failure then
+					request.failure(self.HTTP_REQUEST_FAILED)
```

Comment:
Will we ever be able to return more failure info, and if not (for the forseeable future at least), why have a status code? It could easily be introduced in the future if it's needed, at the moment it's unnecessary.